src/Foundatio.Storage.SshNet/Storage/SshNetFileStorage.cs

```diff
@@ -13,384 +14,432 @@
         }
+    }
+
+    public Task<bool> ExistsAsync(string path) {
+        if (String.IsNullOrEmpty(path))
+            throw new ArgumentNullException(nameof(path));
 
-        public Task<bool> ExistsAsync(string path) {
-            if (String.IsNullOrEmpty(path))
-                throw new ArgumentNullException(nameof(path));
+        EnsureClientConnected();
+        
+        string normalizedPath = NormalizePath(path);
+        _logger.LogTrace("Checking if {Path} exists", normalizedPath);
+        return Task.FromResult(_client.Exists(normalizedPath));
+    }
 
-            EnsureClientConnected();
-            return Task.FromResult(_client.Exists(NormalizePath(path)));
+    public async Task<bool> SaveFileAsync(string path, Stream stream, CancellationToken cancellationToken = default) {
+        if (String.IsNullOrEmpty(path))
+            throw new ArgumentNullException(nameof(path));
+
+        if (stream == null)
+            throw new ArgumentNullException(nameof(stream));
+        
+        string normalizedPath = NormalizePath(path);
+        _logger.LogTrace("Saving {Path}", normalizedPath);
+        
+        EnsureClientConnected();
+
+        try {
+            await Task.Factory.FromAsync(_client.BeginUploadFile(stream, normalizedPath, null, null), _client.EndUploadFile).AnyContext();
+        } catch (SftpPathNotFoundException) {
+            _logger.LogDebug("Error saving {Path}: Attempting to create directory", normalizedPath);
+            CreateDirectory(normalizedPath);
+            
+            _logger.LogTrace("Saving {Path}", normalizedPath);
+            await Task.Factory.FromAsync(_client.BeginUploadFile(stream, normalizedPath, null, null), _client.EndUploadFile).AnyContext();
         }
 
-        public async Task<bool> SaveFileAsync(string path, Stream stream, CancellationToken cancellationToken = default) {
-            if (String.IsNullOrEmpty(path))
-                throw new ArgumentNullException(nameof(path));
+        return true;
+    }
 
-            if (stream == null)
-                throw new ArgumentNullException(nameof(stream));
+    public Task<bool> RenameFileAsync(string path, string newPath, CancellationToken cancellationToken = default) {
+        if (String.IsNullOrEmpty(path))
+            throw new ArgumentNullException(nameof(path));
+        if (String.IsNullOrEmpty(newPath))
+            throw new ArgumentNullException(nameof(newPath));
+
+        string normalizedPath = NormalizePath(path);
+        string normalizedNewPath = NormalizePath(newPath);
+        _logger.LogTrace("Renaming {Path} to {NewPath}", normalizedPath, normalizedNewPath);
+        EnsureClientConnected();
+
+        try {
+            _client.RenameFile(normalizedPath, normalizedNewPath, true);
+        } catch (SftpPathNotFoundException) {
+            _logger.LogDebug("Error renaming {Path} to {NewPath}: Attempting to create directory", normalizedPath, normalizedNewPath);
```

Comment:
Why wouldn't this be logged as an error?

Reply:
Because it can be handled and it's probably a common thing to rename a file across folders.